src/main/java/sparksoniq/jsoniq/runtime/iterator/RuntimeIterator.java

```diff
@@ -132,11 +132,15 @@ public IteratorMetadata getMetadata() {
     protected <T extends Item, E extends SparksoniqRuntimeException> T getSingleItemOfTypeFromIterator(RuntimeIterator iterator,
                                                                                                        Class<T> type, E nonAtomicException) {
         iterator.open(_currentDynamicContext);
-        Item result = iterator.next();
-        if (iterator.hasNext())
-            throw nonAtomicException;
+        Item result = null;
+        if(iterator.hasNext()) {
+            result = iterator.next();
+            if (iterator.hasNext()) {
+                throw nonAtomicException;
+            }
+        }
         iterator.close();
-        if (!(type.isInstance(result)))
+        if (result != null && !(type.isInstance(result)))
```

Comment:
Yes, I think that this is fine this way. This function is purely an internal helper that isn't exposed to the outside world, so you can use the Java null as a special return value that says "there is no such single item".